db/commitlog/commitlog.cc

```diff
@@ -1392,83 +1393,84 @@ void db::commitlog::segment_manager::flush_segments(uint64_t size_to_remove) {
 
             // would be super nice if we just could mmap(/dev/zero) and do sendto
             // instead of this, but for now we must do explicit buffer writes.
-            fut = fsiz.then([f, this, filename](uint64_t existing_size) mutable {
-                // if recycled (or from last run), we might have either truncated smaller or written it 
-                // (slightly) larger due to final zeroing of file
-                if (existing_size > max_size) {
-                    return f.truncate(max_size);
-                } else if (existing_size == max_size) {
-                    return make_ready_future<>();
-                }
-                
+
+            // if recycled (or from last run), we might have either truncated smaller or written it 
+            // (slightly) larger due to final zeroing of file
+            if (existing_size > max_size) {
+                co_await f.truncate(max_size);
+            } else if (existing_size < max_size) {
                 totals.total_size_on_disk += (max_size - existing_size);
 
                 clogger.trace("Pre-writing {} of {} KB to segment {}", (max_size - existing_size)/1024, max_size/1024, filename);
-                return f.allocate(existing_size, max_size - existing_size).then([this, existing_size, f]() mutable {
-                    static constexpr size_t buf_size = 4 * segment::alignment;
-                    size_t zerofill_size = max_size - align_down(existing_size, segment::alignment);
-                    return do_with(allocate_single_buffer(buf_size), zerofill_size, [this, f](temporary_buffer<char>& buf, uint64_t& rem) mutable {
-                        std::fill(buf.get_write(), buf.get_write() + buf.size(), 0);
-                        return repeat([this, f, &rem, &buf]() mutable {
-                            if (rem == 0) {
-                                return make_ready_future<stop_iteration>(stop_iteration::yes);
-                            }
-                            static constexpr size_t max_write = 128 * 1024;
-                            auto n = std::min(max_write / buf_size, 1 + rem / buf_size);
-
-                            std::vector<iovec> v;
-                            v.reserve(n);
-                            size_t m = 0;
-                            while (m < rem && n < max_write) {
-                                auto s = std::min(rem - m, buf_size);
-                                v.emplace_back(iovec{ buf.get_write(), s});
-                                m += s;
-                            }
-                            return f.dma_write(max_size - rem, std::move(v), service::get_local_commitlog_priority()).then([&rem](size_t s) mutable {
-                                rem -= s;
-                                return stop_iteration::no;
-                            });
-                        });
-                    });
-                });
-            });
+
+                // re-open without o_dsync for pre-alloc
```

Comment:
I suggest a more detailed explanation why it is important to re-open it without o-dsync (something from the rationale given in the ticket would do).